rusk/src/lib/chain/rusk.rs

```diff
@@ -495,15 +513,48 @@ fn execute(
     session: &mut Session,
     tx: &PhoenixTransaction,
 ) -> Result<CallReceipt<Result<Vec<u8>, ContractError>>, PiecrustError> {
+    let tx_stripped = strip_off_bytecode(tx);
     // Spend the inputs and execute the call. If this errors the transaction is
     // unspendable.
     let mut receipt = session.call::<_, Result<Vec<u8>, ContractError>>(
         TRANSFER_CONTRACT,
         "spend_and_execute",
-        tx,
+        tx_stripped.as_ref().unwrap_or(tx),
         tx.payload().fee.gas_limit,
     )?;
 
+    // Deploy if this is a deployment transaction
+    if let Some(deploy) = tx.payload().contract_deploy() {
+        if receipt.data.is_ok() {
+            let bytecode_charge =
+                deploy.bytecode.bytes.len() as u64 * GAS_PER_DEPLOY_BYTE;
+            let min_gas_limit = receipt.gas_spent + bytecode_charge;
+            let hash = blake3::hash(deploy.bytecode.bytes.as_slice());
+            if tx.payload().fee.gas_limit < min_gas_limit {
+                receipt.data = Err(OutOfGas);
+            } else if hash != deploy.bytecode.hash {
+                receipt.data = Err(Panic("failed bytecode hash check".into()))
+            } else {
+                let result = session.deploy_raw(
+                    None,
+                    deploy.bytecode.bytes.as_slice(),
+                    deploy.constructor_args.clone(),
+                    deploy.owner.clone(),
+                    tx.payload().fee.gas_limit - receipt.gas_spent,
+                );
+                match result {
+                    Ok(_) => {
+                        receipt.gas_spent += bytecode_charge;
+                    }
+                    Err(err) => {
+                        info!("Tx caused deployment error {err:?}");
+                        receipt.data = Err(Panic("failed deployment".into()))
+                    }
+                }
+            }
+        }
+    };
```

Comment:
What do we think of doing this *before* `spend_and_execute`? At least we should make an effort to see if the transaction has enough gas to fuel this in the first place. Perhaps we could even subtract the gas from the gas given to `spend_and_execute`.

Reply:
Before `spend_and_execute` I could check checksum and enough gas for bytecode charge, yet I cannot penalize the caller without `spend_and_execute`, I could only discard the TX, and in the current code I really penalize the caller - she loses the entire gas limit if any of these conditions occur. I think such penalties in the context of deployment make sense, as they discourage spamming. Do you think we should be more lenient in these cases? I'd think definitely not in case of bad checksum, so it is only the bytecode charge that remains.